### PR TITLE
Angular ify

### PR DESCRIPTION
**Pull request by kimberlythegeek**
_Note: There are still style changes I'd like to make, but this branch is functionally on par with the current one. Default view of issues is now set to collapsed, and repos will not show if they contain no issues meeting the filter criteria._

__Notes on current code base:__
<a href="http://docs.google.com/gview?url=https://github.com/mozilla/fxtest-dashboard/files/593192/fxtest-dashboard-angularify.pdf&embedded=true">View on Google Docs</a>
[Download PDF](https://github.com/mozilla/fxtest-dashboard/files/593192/fxtest-dashboard-angularify.pdf)

_I welcome your feedback._

 _/kimberlythegeek_

___

# Notes on App Refactor

## Contents

- [config.json](#configjson)
- [main.js](#mainjs)
  - [IIFE](#iife)
  - ['use strict'](#use-strict)
  - [Instantiation](#instantiation)
  - [Dependency Injection](#dependency-injection)
  - [DOM Manipulation](#dom-manipulation)
  - [jQuery](#jquery)
  - [$routeProvider](#routeprovider)
- [aggregator.js](#aggregatorjs)
- [Issues Addressed](#issues-addressed)
    - [#52](#issue-52)
    - [#60](#issue-60)
- [Mozilla Style Guides](#mozilla-style-guides)

___

# config.json

```
{
  "repos": [
    "https://github.com/mozilla/Addon-Tests",
    "https://github.com/mozilla/bidpom",
    "https://github.com/mozilla/fxapom",
    "https://github.com/mozilla/fxtestbot",
    "https://github.com/mozilla/hello-tests",
    "https://github.com/mozilla/moz-grid-config",
    "https://github.com/mozilla/mozillians-tests",

    .....

  ],
  "api_token": "692aadd2c9789a337c4d3544787f101ad8c35bc7"
}
```

Rather than an array of URLs, I restructed the `repos` array to contain Object Literals.
Including `https://github.com/` in every entry is redundant, becuase it will always be the same.

This also adds the redundant task of extracting the `owner` and `name` of the repository from each entry.

The new structure is as follows:

```
{
  "repos": [
      {
          "owner": "mozilla",
          "name": "Addon-Tests"
      },
      {
          "owner": "mozilla",
          "name": "bidpom"
      },
      {
          "owner": "mozilla",
          "name": "fxapom"
      },
      {
          "owner": "mozilla",
          "name": "fxtest-dashboard"
      },

      .....

      ],
      "api_token": "692aadd2c9789a337c4d3544787f101ad8c35bc7"
}
```

Within the FeedController, the `"repos"` array also becomes the data structure to hold the array of issues for each repository, so each repo object also contains its own issues. The original app contatined two separate arrays.

The `issues` array  is _only_ added to each `repo` object _if open issues exist_.

___

# main.js

### IIFE
Wrapping the code in an [IIFE](https://en.wikipedia.org/wiki/Immediately-invoked_function_expression)
`()();` prevents declared variables and functions from overflowing out of the scope.

### 'use strict'
Specifying `'use strict';` will throw errors that will otherwise have been ignored.

### "Modularized"
I moved all module artifacts&#8212;controllers, directives, services, etc.&#8212;into separate files.
This both improves organization, and improves code reusability.

All module artifact files are found in `src/dashboard/`, following the pattern of `[name].controller.js`, `[name].filter.js`, `[name].template.html`, etc.

### Dependency Injection
The app is currently lacking dependency injection on both controllers&#8212;
`FeedController.$inject = ['$scope', 'GitHubIssueFeedService']` which protects the code from minification.

### DOM Manipulation

Throughout the app, there are several places where the code is directly manipulating the DOM, as with jQuery.
Being an [MVVM Framework](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93viewmodel), this is not
recommended, as this can be done dynamically with AngularJS.

### jQuery

Much of the logic in the app is done via jQuery, not Angular. There are much simpler and more efficent
ways to implement the functionality using controllers, services, directives, etc.

### $routeProvider

Using `$routeProvider` (main.js line 15) is unnecessary for this app. Instead of
having two separate pages for Issues and Pull Requests, (since pull requests are issues),
you can use the same template and filter for pull requests when viewing the Pull Requests Page.

### Redundant Code
There are many places in the app where redundant code is used.

- IssuesController & PullRequestsController
  - These controllers are virtually the same, except the latter filters the data down to just pull requests.
  - The logic of showing only pull requests can be done using Angular filters and `ng-if` statements.
  - The controllers also set up redundant data structures, when all of the data
  can be accessed within the objects retrieved from the API

- Within the Controllers, there is a `forEach` loop (lines 65 & 165) that sets up
the display properties (should issue/repository be shown or not?). This can be done
automatically and dynamically with AngularJS.

- Line 85: `$scope.toggleRepo` is a function to show/hide repository when arrow
is clicked. Can be done using Angular's `ng-click` and `ng-show`.

- Line 89: `$scope.showHideIssues` is a function that loops through all of the data
every time something changes to determine whether or not it should be shown. All
it is unnecessary, the logic can be implemented using Angular filters and directives.

### Custom Directives

I have created several custom directives for blocks of code that are reused, i.e. `<github-repository>`
 and `<filter-list>`. Directives are comprised of a declaration file `[name].directive.js`, and a template file `[name].template.html`.

### Custom Service

I creatd the `issueFeed` service to handle JSON calls, passing the data to the `FeedController`.
___

# aggregator.js

This file handles most of the logic. All of the code in this file can be
incorporated into main.js within controllers, services, filters, etc. It makes
the API calls, and declares the data structures for the repositories and issues.

It is also guilty of using jQuery for almost all of the logic. Everything in
I have broken down into functions within GitHubIssueFeedService and FeedController.

Instead of using jQuery's getJSON method, we can utilize Angular's `$http` service.
This also eliminates the need for the paramaters in our API URL, as the $http service
can take `params: {}` and parse them appropriately.
___

# Issues Addressed

### Issue #52

While creating filter functionality, I decided to also address [Issue #52](https://github.com/mozilla/fxtest-dashboard/issues/52) to avoid rewriting the filters later.

Is Assigned filter is now a checkbox. Uncheck All links have been added.

### Issue #60

Also addressing this issue to avoid later rewrite. Labels are now pulled from the found open issues  [Issue #60](https://github.com/mozilla/fxtest-dashboard/issues/60)

___

# Mozilla Style Guides

I have begun editing all of the CSS, HTML, and JS to follow Mozilla's
[style guides](http://mozweb.readthedocs.io/en/latest/reference/index.html).

I also add comments next to closing tags in HTML, as a preference, to make it
easier on both myself and anyone else who may read the code after me.

### CODE REVIEW

**Comment by davehunt** [comment]
Wow @kimberlythegeek, there's a lot in here! It appears there are some conflicts though, perhaps from some of your other patches that have landed. Could you rebase this branch from upstream, or merge from upstream so that the conflicts are resolved? Once that's done I'll take it for a spin.

**Comment by kimberlythegeek** [comment]
Okay, so I've never done a rebase before&#8212;I _think_ I did it correctly, let me know if the conflicts still exist.

Thanks!

File: issue-notes.md [hunk 1]
```diff
@@ -0,0 +1,190 @@
+# Notes on App Refactor
```
Comment (davehunt) [inline issue-notes.md]: Did you mean to commit this file?

**Review by davehunt — CHANGES_REQUESTED** [review]
I'm not able to review the code in detail because I'm not familiar enough with angular, but the refactoring does make a lot of sense. I've also tested the dashboard locally and it works well. I noticed a few of JavaScript warnings in the console, and we should remove `issue-notes.md` from the patch. With these addressed, I'm happy to squash and merge.

**Pull request closed without merging**